docs/rfc/0004_component_groups_scale_view.md

```diff
@@ -0,0 +1,300 @@
+    component: frontend     # One of meta, frontend, compute, and compactor.
+  replicas: 10              # Target replicas. It will be synced from the RisingWave object when the controller grabs the lock.
+  # Strict verification mode on replicas:
+  #   1. If enabled, the validating webhook will reject any invalid changes on .spec.replicas (e.g., exceeding the min/max range)
+  #   2. If disabled, the mutating webhook will adjust the .spec.replicas to be enclosed by the range.
+  # Serialized label selector. Would be set by the webhook.
+  strict: false
+  labelSelector: "risingwave/name=hello,risingwave/component=frontend,risingwave/group in (enduring,spot)"
+  scalePolicy: # An array of groups and the policies for scale, 
+  # optional and empty means the default group with the default policy.
+  # The order here indicates the priorities.
+
+  - group: 'enduring'       # Target group name
+    # 0-10, optional. The groups will be sorted by the priority and the current replicas. 
+    # The higher it is, the more replicas of the target group will be considered kept, i.e. scale out first, scale in last.
+    priority: 1
+    replicas: 2             # Controlled field, maintained by the mutating webhook. Current desired replicas.
+    constraints:
+      min: 1              # >= 0, optional, default is 0
+      max: 2              # >= minimum, optional, default is 1000 (indicates infinite because 1000 Pods are many enough to be considered as infinite).
+  - group: 'spot'
+    priority: 0
+    replicas: 3
+status:
+  targetRef: # It has value only if the scale view has grabbed the lock on the target object.
+    uid:                    # UID of the target object.           
+  replicas: 10             # Current replicas of target groups. Maintained by the controller.
+```
+
+### Priority and Replica Constraints
+
+When updating the scale subresource, it's updating the `.spec.replicas` field above. The mutating webhook will then work
+to distribute the replicas of `.spec.replicas` to the replicas under the `scalePolicy`, following the rules below:
+
+- The replicas in each group must be confined within the constraints.
+- The groups with higher priorities always consume replicas first until it reaches the max boundary. E.g.,
+  when `.spec.replicas` is 4, the 'enduring' group (with priority 1, larger than the 'spot' group's) will have 2 (which
+  is the max) replicas, and then the left is assigned to the 'spot' group.
+- The groups with the same priorities tend to have similar replicas, within the constraints. Which is to say, the target
+  is to minimize the variance.
+
+## `RisingWave` CRD and Webhooks
+
+```yaml
+apiVersion: risingwave.risingwavelabs.com/v1alpha1
+kind: RisingWave
+metadata:
+  name: hello
+# ....
+status:
+  scaleViews: # An array of scale view summaries. Maintained by the RisingWaveScaleView controller.
+  - name: hello-frontend-scale-view # Name of the RisingWaveScaleView object.
```

Comment:
Is it correct to assume scaleview is only used by frontend, metanode, and compactor?

Reply:
If you intentionally left the compute node out, then no.